### DISCUSSION TITLE
Allow for chaining a `title()` method on the `render()` method

### DISCUSSION THREAD

**Discussion by jeroenvanrensen** [livewire/livewire#4195] [3. Feature ideas]
Hey everyone,

Say, you've got this component:

```php
public function render()
{
    return view('auth.login')->layoutData(['title' => 'Login']);
}
```

Wouldn't it be nice if you could write it like so:

```php
public function render()
{
    return view('auth.login')->title('Login');
}
```

So we keep the `layoutData()` method, but specifically for the title you can use the `title()` method.

I'm curious to hear your thoughts about it!

Thanks! Jeroen

**Comment by zepfietje** [comment]
Hey Jeroen! I've also been thinking about reducing boilerplate for setting layout data (also see discussion in https://github.com/livewire/livewire/pull/2736).

However, I'm not sure if it's a good idea to add specific methods for layout data like `title` (or `description` for example). Then you'd basically want to support any layout variable name, which defeats the purpose.

What I'd been thinking about is something a little different. Since you probably specify layout data like `title` in every full-page Livewire component, it'd be nice to not have to specify the view in the render method (or the render method entirely). Of course this only applies if you stick to standard component naming and directories. What I envision is a component method like `layout` or `layoutData` which allows you to specify layout data like `title` without having to include any other boilerplate code.

Let me know your thoughts!

**Reply by joshhanley** [reply]
@jeroenvanrensen yeah I agree with @zepfietje using named methods isn't great when not everyone will have a title method in their layout (or might have other properties).

But I do agree there could be a nicer way to handle it.

@zepfietje yeah that idea of a dedicated layout method has been discussed, see #3138, and I think could be the way to go.

Would be nice to be able to specify layout data without actually having to specify the layout, that way if you are just using the default `app.blade.php` we don't need to add that each time.

**Reply by zepfietje** [reply]
https://github.com/livewire/livewire/discussions/3138 is related indeed. I feel like having a `layout()` method is even better than `layoutData()` (to prevent separating every little bit into a different method). Probably even find it a little neater than using `->layout()` in the `render()` method.

A solution could be a `layout` method (or property) that could specify a layout view or leave it out and just use the default `layouts.app`. To build on the suggestion of @calebporzio:

```php
class Foo extends Component
{
    // You can specify all these parameters
    protected array $layout = [
        'view' => 'layouts.dashboard',
        'slot' => 'body',
        'data' => [
            'title' => 'Page title',
        ],
    ];

    // Or just specify data
    protected array $layout = [
        'data' => [
            'title' => 'Page title',
        ],
    ];

    // Or use the method
    protected function layout(): array
    {
        return [
            'data' => [
                'title' => 'Page title',
            ],
        ];
    }
}
```

Similar thing would be added for `extends` if that's what you use instead of `layout`.
Separating layout configuration from the `render()` method allows you to be more flexible (as you are no longer tied to specifying the component view in addition to layout config). This way it's super easy to add a `title()` method to your own project like @jeroenvanrensen suggested.

It would be nice if we could get this feature implemented. Really feel like this is a small but nice addition to Livewire!

**Reply by joshhanley** [reply]
@zepfietje yep that would work!

If you feel up to it, I'm sure a PR for it would be welcome 🙂

**Reply by zepfietje** [reply]
Had a quick look, it shouldn't be too difficult. Will give it a shot soon!

**Reply by jeroenvanrensen** [reply]
@zepfietje I have been thinking about what you said. Instead of what you're suggesting, maybe something is possible with macro's or PHP's magic `__call()` method? I don't know enough about Livewire's internals to know if it is possible, but it seems like a great alternative.

**Reply by zepfietje** [reply]
@jeroenvanrensen you can probably implement the `title()` method you suggested using a macro indeed. However, as discussed above, it's better to go with a generic solution which allows you to build opinionated behavior on top of it.

**Reply by kokoshneta** [reply]
Did anything ever come of this? As far as I can tell, per-component layout configuration is still not possible, which makes me sad…
